src/timely-util/src/order.rs

```diff
@@ -553,3 +554,194 @@ mod test {
         assert_eq!(part_summary1.results_in(&ts4), None);
     }
 }
+
+pub mod hybrid {
```

Comment:
@bkirwi said there might be one other impl that was wrong, were they exactly the same other than `Refines`?

Reply:
talked offline, we will ensure the timely pr merges

Reply:
I continue to believe that [this](https://github.com/MaterializeInc/materialize/pull/19758/files#diff-cb9970c8cd8d0ba36ab1b78e30305c768908b64035b48cf76d76e038e375fc1fR636-R662) is the right summary implementation for lexicographically-ordered timestamps, so it would be nice to roll that in if we're discarding this code. But it sounded like @petrosagg disagreed?

Reply:
@bkirwi we can fix this in the timely pr as well, as that lexographic ts is not used anywhere I don't think...

Reply:
The thing is that there is no such thing as "right" summary, so it's impossible to agree or disagree :D Summaries are there to summarize what happens to timestamps when they cross certain paths, and so the only requirement for the summary type would be to be expressive enough for the paths you wish to summarize.

If you never intend to iterate using this timestamp you might as well define its summary to be `()` and then the only paths that can ever exist are paths that leave the timestamp alone. That's a totally fine choice.

Or you may do intend to iterate using lexical timestamp (although is problematic for other reasons[1]) and in that case you'd want summaries to be expressive enough to be able to encode an "increase the second coordinate" action. For that purpose both the implementation in `timely` and this one are sufficient.

You can probably construct a dataflow that really needs to have paths that timestamps get transformed according to this implementation and not the other one but I don't really understand what that dataflow would be useful for or what it would do.

All that is to say, for all practical purposes (i.e prioritized iteration) the two summary descriptions are equivalent so my argument is just "let's have less code to maintain".

[1] https://link.springer.com/content/pdf/10.1007/978-3-662-46678-0_5.pdf (search for "via lexicographic products")

Reply:
> The thing is that there is no such thing as "right" summary [...]

I mean, sure. But timely gets along fine with the idea that there is a canonical summary type for a particular timestamp, because for many timestamp types there is a reasonable best choice for that timestamp's behaviour and semantics. (The summary type for `Product` seems to be optimal in that way... while there are other valid ones, product's summary type seems to be the most useful notion of path that satisfies the restrictions.) 

> Or you may do intend to iterate using lexical timestamp [...] in that case you'd want summaries to be expressive enough to be able to encode an "increase the second coordinate" action. For that purpose both the implementation in timely and this one are sufficient.

Right. But the current semantics for tuples given nonsensical results when increasing the first coordinate, which is also a useful thing to express in contexts like this one.

> All that is to say, for all practical purposes (i.e prioritized iteration) the two summary descriptions are equivalent so my argument is just "let's have less code to maintain".

Assuming we're still talking about whether or not we try and change the upstream summary implementation, I don't follow this argument.

Anyways, I don't think this is worth being a stickler for, since if / when we run into summary pain we can make the change later. But I do think we're throwing a baby out with some bathwater here.

Reply:
> Assuming we're still talking about whether or not we try and change the upstream summary implementation, I don't follow this argument.

No I was not arguing about changing upstream, that seems like a fine thing to do although I would expect that you'd need to come up with a motivating example. I was arguing that since the upstream definition covers all our needs, and we don't have any case where this alternative summary implementation is needed we can just use the upstream definition and have less code to think about in `timely-util`.

> But the current semantics for tuples given nonsensical results when increasing the first coordinate.

I didn't follow that, why is it nonsensical to increase the first coordinate while keeping the second one intact? I can't come up with a dataflow that needs it but I also can't come up with a dataflow that needs to increase the first coordinate and reset the second one.